refactor(all): Convert to TS WIP

bigopon opened the pull request:
I think we should convert this repo to TS, as its current state is too error prawn

After this we should be able to iterate faster and safer.

@davismj @EisenbergEffect @fkleuver 

TODO: redo `gulp doc` task

@EisenbergEffect What gulp tasks do I have to do to convert this to rollup + TS ?

EisenbergEffect commented:
@bigopon I don't know anything about the roll-up work. Regarding the doc task, we don't need that in the repo anymore. It's not actually used to generate the docs. It's still useful for non-TS repos because I use that to validate that the output d.ts files are actually correct. For TS, they're guaranteed to be correct. However, the real API generation is still a problem.That's handled by the site-gnerator. It doesn't work for native TS libraries currently. So, we can't release this in TS without fixing that. FYI This is why there's no API documentation for Validation either.

bigopon commented:
@EisenbergEffect @davismj It's in good state now. All tests passed. I have to add this [line](https://github.com/aurelia/router/pull/615/files#diff-d8377f95ecc9f3984ff84c55ecb6a083R163) because the way code is compiled for test (to raw es2015, so this [line](https://github.com/aurelia/router/pull/615/files#diff-52fd7ff9c4a70144c285b45938c2af80R140) will fail without `lifecycleArgs` ). Beside that and some extra typings, all remain the same

EisenbergEffect commented on package.json:
```diff
@@ -62,6 +63,8 @@
   },
   "devDependencies": {
+    "@types/jasmine": "^2.8.8",
+    "aurelia-templating": "^1.8.2",
```
What's the need for adding the templating library here?

bigopon replied:
@EisenbergEffect It's for `import { CompositionContext } from 'aurelia-templating'` for better intellisense.

EisenbergEffect commented on src/interfaces.ts:
```diff
@@ -1,11 +1,35 @@
+}
+
+/**@internal */
+declare module 'aurelia-templating' {
```
I understand why the above dependency was added. Bummer. This is a breaking change probably. We can live with it though.

bigopon replied:
Maybe not ... ? if we use it in this way, the code we deliver won't get affected ? Or we can just retype `CompositionContext` in this lib

bigopon replied:
Retype `CompositionContext` or a part of it only

EisenbergEffect commented on src/interfaces.ts:
```diff
@@ -106,93 +130,93 @@ interface RouteConfig {
   * a view model for different routes. Available values are 'replace' and 'invoke-lifecycle'.
   */
-  determineActivationStrategy: (params: any, routeConfig: RouteConfig, navigationInstruction: NavigationInstruction) => 'no-change' | 'invoke-lifecycle'| 'replace';
+  determineActivationStrategy(params: any, routeConfig: RouteConfig, navigationInstruction: NavigationInstruction): 'no-change' | 'invoke-lifecycle' | 'replace';
```
Should the return type also use a `keyof`?

bigopon replied:
yes, I missed this. Should be `ActivationStrategy[keyof ActivationStrategy]`

EisenbergEffect commented on src/interfaces.ts:
```diff
@@ -106,93 +130,93 @@ interface RouteConfig {
 * An optional interface describing the available activation strategies.
 */
-interface ActivationStrategy {
+export interface ActivationStrategy {
```
Should we convert this to an enum?

bigopon replied:
I was doing so, that's why you see `,` instead of `;` in its fields. I'm not sure about it, because it's already exported, and no1 uses this interface, only the `export const activationStrategy`

fkleuver replied:
This interface never made any sense to begin with because it describes an object containing activationStrategies, not an actual activationStrategy (which is just a plain string).
*This* would make sense: `export type ActivationStrategy = 'no-change' | 'invoke-lifecycle' | 'replace'`

What this interface *could* be useful as, is `HasActivationStrategy`. E.g. you could make a ViewModel implement it. But then still it doesn't make sense because it doesn't have a property `activationStrategy: 'no-change' | 'invoke-lifecycle' | 'replace'` (like it should).

Since the object (or enum, they're the same thing) are both capable of describing themselves, this interface is unquestionably redundant.

StrahilKazlachev replied:
@fkleuver Actually the `interface ActivationStrategy` should describe a `const` similar to the `bindingMode` one. Though not sure where it is 😅

davismj replied:
@StrahilKazlachev here https://github.com/aurelia/router/blob/master/src/navigation-plan.js#L6

@bigopon Sounds like we can remove this completely, right?

bigopon replied:
Yes, probably. I added a dedicated interface
https://github.com/aurelia/router/blob/a0135eba04cdf222795fff2c58cedd6a1987969e/src/interfaces.ts#L210

davismj replied:
@bigopon Should this be removed?

bigopon commented:
@davismj @EisenbergEffect I removed the `aurelia-templating` dependency, and added a new interface [here]( https://github.com/aurelia/router/pull/615/commits/8e5d11bcc215cd5e3f68be2a068417a44eab4522#diff-a84f60e04d03dd963458bd8a12f9ac20R240) (It's compatible with `CompositionContext`)

For changes like this https://github.com/aurelia/router/pull/615/files#diff-52fd7ff9c4a70144c285b45938c2af80R9 , it's a semi breaking change: those classes aren't expected to be used outside of `router`, but if folks complain, we can change it to something like this:

```ts
  run(navigationInstruction: NavigationInstruction, next: Partial<Next>)
```

The `Partial` will make it compatible with `Function`

bigopon commented:
@rluba thanks for #602 , I stole the idea and added it in. Code doesn't look very nice but it somewhat does the job. If anyone has an idea to improve it, please send help 😄  to: https://github.com/aurelia/router/pull/615/files#diff-2624298877ec987a3e49f3a716b66329

@EisenbergEffect @davismj I also added new build / dev script in the README

fkleuver commented on karma.conf.js:
```diff
@@ -1,81 +1,50 @@
+            loader: "ts-loader",
+            exclude: /node_modules/,
+            options: {
+              transpileOnly: true
```
I typically turn this on only when a code base is volatile and I just want tests to run even with typing errors. It also speeds up the compilation to have this on.
You might want to remove this or make it parameterized as a measure of robustness for this mature library. It's probably better to see typing errors during testing as well.

bigopon replied:
Fixed, thanks !

bigopon commented:
@davismj Yes, but the classes exported (which have `Function` changed to `Next`) won't normally be used, which makes it a minor issue, so I think it's ok to process. But we have options to handle it.

davismj commented on src/activation.ts:
```diff
@@ -55,14 +64,18 @@ function processDeactivatable(navigationInstruction: NavigationInstruction, call
+function findDeactivatable(
+  plan: Record<string, ViewPortInstruction>,
+  callbackName: string,
+  list: Object[] = []
```
`IActivatableInfo[]`

davismj commented on src/activation.ts:
```diff
@@ -132,36 +150,57 @@ function processActivatable(navigationInstruction: NavigationInstruction, callba
+      let viewPortInstruction = navigationInstruction.viewPortInstructions[viewPortName];
+      let viewModel = viewPortInstruction.component.viewModel;
+
+      if (
```
nitpick but this is not any easier to read and takes up more vertical space, which i see as a net negative

bigopon replied:
I see that it's not easier to read, but the current state doesn't seem to help either, so I picked the 80 chars a line rule and go with it. We can have this line as an exception though, what do you think about a tslint file

davismj commented on src/app-router.ts:
```diff
@@ -14,6 +23,15 @@ const logger = LogManager.getLogger('app-router');
   static inject() { return [Container, History, PipelineProvider, EventAggregator]; }
 
+  /**@internal */
+  events: EventAggregator;
```
not internal, this is part of the public api

davismj commented on src/app-router.ts:
```diff
@@ -1,10 +1,19 @@
+import { PipelineProvider } from './pipeline-provider';
+import { isNavigationCommand } from './navigation-commands';
+import { EventAggregator } from 'aurelia-event-aggregator';
+import { NavigationInstruction } from '.';
```
./navigation-instruction?

bigopon replied:
they was all rexported from index, I think that's why I got this automatically completed from the wrong module

davismj commented on src/app-router.ts:
```diff
@@ -54,16 +72,18 @@ export class AppRouter extends Router {
-          return this.configure(config => viewModel.configureRouter(config, this))
+          this._resolveConfiguredPromise = () => { };
+          return this
+            // potential error here as callback function should return RouterConfiguration
```
```
            .configure(config => {
              viewModel.configureRouter(config, this);
              return config;
            })
```

bigopon replied:
nice

davismj commented on src/interfaces.ts:
```diff
@@ -201,14 +217,62 @@ interface PipelineStep {
+
+  childRouter?: Router;
+
+  lifecycleArgs?: any[];// [Record<string, any>, RouteConfig, any]
```
[Record<string, string>, RouteConfig, NavigationInstruction]

davismj commented on src/interfaces.ts:
```diff
@@ -201,14 +217,62 @@ interface PipelineStep {
+  process(viewportInstruction: ViewPortInstruction, waitToSwap?: boolean): void;
+}
+
+export interface ViewPortInstruction {
```
Having name be the only required part doesn't make sense to me. I don't think name is used anywhere at all. I believe strategy, config, moduleId, component, and lifecycleArgs are all required.

bigopon replied:
I needed to make it partial to get it to work as easily as possible. Will improve the typing here

davismj commented on src/nav-model.ts:
```diff
@@ -38,6 +41,9 @@ export class NavModel {
   */
   router: Router;
 
+  /**@internal */
```
I'm not sure if this should be internalized. I could see someone depending on it?

bigopon replied:
Mainly added this in typing for internal usage and it does not exist in our current distribution, I think we may not need to worry about it.

davismj replied:
Order is part of the public api. See here for example: https://aurelia.io/docs/routing/configuration#basic-configuration

davismj commented on src/navigation-commands.ts:
```diff
@@ -13,19 +27,28 @@ interface NavigationCommand {
    * @param options The navigation options.
    */
-  constructor(url: string, options: any = {}) {
+  constructor(url: string, options: Partial<NavigationOptions> = {}) {
```
Why Partial?

bigopon replied:
Here is what I need to work a bit more, `Redirect` is a commonly used class, and the existing typing for 2nd param is any. I need to make the new typing strong, while ensuring backward compat. I thought `Partial` would do the job but it seems I will need to do something else

davismj commented on src/navigation-instruction.ts:
```diff
@@ -1,13 +1,18 @@
+  queryString?: string;
+  params?: Record<string, any>;
+  queryParams?: Record<string, any>;
+  config?: Partial<RouteConfig>;
```
Why partial? Why optional?

davismj commented on src/navigation-instruction.ts:
```diff
@@ -1,13 +1,18 @@
+  config?: Partial<RouteConfig>;
+  parentInstruction?: NavigationInstruction;
+  previousInstruction?: NavigationInstruction;
+  router?: Router;
```
why optional

davismj commented on src/navigation-instruction.ts:
```diff
@@ -1,13 +1,18 @@
+import { activationStrategy } from './navigation-plan';
+
+export interface NavigationInstructionInit {
+  fragment?: string;
```
why optional?

davismj commented on src/navigation-instruction.ts:
```diff
@@ -121,11 +133,11 @@ export class NavigationInstruction {
-    let viewportInstruction = this.viewPortInstructions[viewPortName] = {
-      name: viewPortName,
-      strategy: strategy,
+  addViewPortInstruction(viewPortNameOrInstruction: string, strategy: string, moduleId: string, component: any): ViewPortInstruction {
```
why rename viewPortName. It is still typed as a string, and I don't know anywhere that passes in a full instruction, nor is there any handling for that.

bigopon replied:
I was basing this on an attempt of overloading `addViewPortInstruction` and deprecate the current one for static router. Forgot to revert the name. for `strategy`, I wanted to but wasn't sure at first as it was a break change, but again, I think it's semi private and internal, probably fine to convert to keyof

davismj commented on src/navigation-instruction.ts:
```diff
@@ -121,11 +133,11 @@ export class NavigationInstruction {
+    const config = Object.assign({}, this.lifecycleArgs[1], { currentViewPort: viewPortNameOrInstruction });
+    const viewportInstruction = this.viewPortInstructions[viewPortNameOrInstruction] = {
+      name: viewPortNameOrInstruction,
+      strategy: strategy as any,
```
if strategy argument is typed correctly this can be just strategy.

davismj commented on src/navigation-instruction.ts:
```diff
@@ -204,7 +217,7 @@ export class NavigationInstruction {
 
     for (let viewPortName in this.viewPortInstructions) {
-      let viewPortInstruction = this.viewPortInstructions[viewPortName];
+      let viewPortInstruction: ViewPortInstruction = this.viewPortInstructions[viewPortName];
```
unnecessary

bigopon replied:
Got the typing in before `this.viewPortInstructions`, Removing next commit

davismj commented on src/navigation-plan.ts:
```diff
@@ -33,7 +41,7 @@ export function _buildNavigationPlan(instruction: NavigationInstruction, forceLi
           // If the param on the redirect points to another param, e.g. { route: first/:this, redirect: second/:this }
           let val = newInstruction.params[key];
-          if (typeof(val) === 'string' && val[0] === ':') {
+          if (typeof (val) === 'string' && val[0] === ':') {
```
nitpick but if you're going to add the space might as well drop the parens

davismj commented on src/pipeline-provider.ts:
```diff
@@ -1,20 +1,29 @@
+  /**@internal */
+  slotName: string;
+  /**@internal */
+  slotAlias: string;
```
optional

davismj commented on src/pipeline-provider.ts:
```diff
@@ -1,20 +1,29 @@
   steps = [];
 
-  constructor(container, name, alias) {
+  constructor(container: Container, name: string, alias: string) {
```
alias?: string

bigopon replied:
yes, I missed this

davismj commented on src/pipeline-provider.ts:
```diff
@@ -29,7 +38,12 @@ class PipelineSlot {
+  /**@internal */
+  container: Container;
+  /**@internal */
+  steps: (Function | PipelineSlot)[];
```
I don't think steps is ever a function.

bigopon replied:
In the constructor of this class, we are adding some step constructors for later instantiation. like this [line](https://github.com/aurelia/router/pull/615/files/f41a488866ce19b1e3e2bc859262665818990519#diff-e0658779b23fb0d91b6c56ce52c62769R53)

davismj replied:
They are all PipelineSteps. The placeholders are also PipelineSteps.

davismj commented on src/pipeline-provider.ts:
```diff
@@ -58,8 +72,9 @@ export class PipelineProvider {
-    return this.steps.find(x => x.slotName === name || x.slotAlias === name);
+  /**@internal */
+  _findStep(name: string): PipelineSlot {
+    return this.steps.find((x: PipelineSlot) => x.slotName === name || x.slotAlias === name) as PipelineSlot;
```
fix this.steps type and you won't need the PipelineSlot typings here

davismj commented on src/pipeline-provider.ts:
```diff
@@ -68,8 +83,8 @@ export class PipelineProvider {
     if (found) {
-      if (!found.steps.includes(step)) { // prevent duplicates
-        found.steps.push(step);
+      if (!(found as PipelineSlot).steps.includes(step)) { // prevent duplicates
```
unneccessary typing

davismj commented on src/pipeline-provider.ts:
```diff
@@ -68,8 +83,8 @@ export class PipelineProvider {
-      if (!found.steps.includes(step)) { // prevent duplicates
-        found.steps.push(step);
+      if (!(found as PipelineSlot).steps.includes(step)) { // prevent duplicates
+        (found as PipelineSlot).steps.push(step);
```
unnecessary typing

bigopon replied:
Was because `found` could be either `PipelineSlot` or `Function` from existing type definition

bigopon replied:
Edit: It was necessary because of [this](https://github.com/aurelia/router/pull/615/files/f41a488866ce19b1e3e2bc859262665818990519#diff-e0658779b23fb0d91b6c56ce52c62769R53)

davismj replied:
I never was able to see what you were linking to. Do you know what its necessary?

bigopon replied:
Somehow the old commits are gone. It's necessary because the way it works with container. If we gives it a Function, it will ask the container to create an instance of it, otherwise, as object, we assume it's an object with `PipelineSlot` shape

davismj replied:
For whatever reason, I was thinking `ActivateNextStep` was an actual step and not the class constructor.

davismj commented on src/route-loading.ts:
```diff
@@ -0,0 +1,145 @@
+  /**@internal */
+  routeLoader: RouteLoader;
+
+  constructor(routeLoader: RouteLoader) {
```
is there an advantage to @internal over 

`constructor(private routeLoader: RouteLoader)`?

bigopon replied:
if we used `private`, the `d.ts` will still have it, while `@internal` strips it entirely from the generated result. It doesn't really matter much unless we have a lot of comments for `private` for better understanding from interfaces

davismj commented on src/route-loading.ts:
```diff
@@ -0,0 +1,145 @@
+  return Promise.all(loadPromises);
+}
+
+interface ILoadingPlan {
```
Maybe shift to interfaces

bigopon replied:
The only function that uses this interface is local to this module so I thought it may as well sit in the same place

davismj commented on src/router-configuration.ts:
```diff
@@ -16,11 +18,14 @@ export class RouterConfiguration {
-  unknownRouteConfig: string|RouteConfig|((instruction: NavigationInstruction) => string|RouteConfig|Promise<string|RouteConfig>);
-  viewPortDefaults: {[name: string]: {moduleId: string|void; [key: string]: any}};
+  unknownRouteConfig: string | RouteConfig | ((instruction: NavigationInstruction) => string | RouteConfig | Promise<string | RouteConfig>);
+  viewPortDefaults: { [name: string]: { moduleId: string | void;[key: string]: any } };
```
nitpick but you lost the space after the semicolon, maybe break this out onto multi lines

davismj commented on src/router-configuration.ts:
```diff
@@ -110,7 +115,7 @@ export class RouterConfiguration {
    * @chainable
    */
-  useViewPortDefaults(viewPortConfig: {[name: string]: {moduleId: string; [key: string]: any}}): RouterConfiguration {
+  useViewPortDefaults(viewPortConfig: { [name: string]: { moduleId: string;[key: string]: any } }): RouterConfiguration {
```
nitpick but lost space after semicolon

bigopon replied:
auto format of vscode x_x

davismj commented on src/router-configuration.ts:
```diff
@@ -193,8 +198,9 @@ export class RouterConfiguration {
-        let {name, step} = pipelineSteps[i];
-        pipelineProvider.addStep(name, step);
+        let { name, step } = pipelineSteps[i];
+        // Potential error, addStep only accepts PipelineStep, but step here may be also a function
```
should never be a function, as far as I know.

bigopon replied:
It's true that it cant be a function here, the existing typing is inaccurate.

davismj commented on src/router.ts:
```diff
@@ -111,6 +143,23 @@ export class Router {
   viewPortDefaults: any = {};
 
+  /**@internal */
+  catchAllHandler: (instruction: NavigationInstruction) => NavigationInstruction | Promise<NavigationInstruction>;
```
not internal

bigopon replied:
@davismj not public either ? I don't see it in the current `aurelia-router.d.ts`

davismj replied:
We were having problems generating the router.d.ts, which probably explains this.

bigopon replied:
I'd vote for it to be like it is atm. We can always change it to public if necessary.

davismj commented on src/router.ts:
```diff
@@ -111,6 +143,23 @@ export class Router {
+  /**@internal */
+  catchAllHandler: (instruction: NavigationInstruction) => NavigationInstruction | Promise<NavigationInstruction>;
+  /**@internal */
+  fallbackRoute: string;
```
not internal

bigopon replied:
Same for this, it's not in the router in our current `aurelia-router.d.ts`

bigopon replied:
I'd vote for it to be like it is atm. We can always change it to public if necessary.

davismj commented on src/interfaces.ts:
```diff
@@ -201,14 +217,62 @@ interface PipelineStep {
+  prevModuleId?: string;
+}
+
+export type NavigationResult = boolean | Promise<PipelineResult> | Promise<PipelineResult | boolean>;
```
doesn't `Promise<PipelineResult | boolean>` imply `Promise<PipelineResult>`?

davismj commented on src/router.ts:
```diff
@@ -191,25 +240,27 @@ export class Router {
 
-    let result = callbackOrConfig;
-    let config;
+    let result: RouterConfiguration = callbackOrConfig as RouterConfiguration;
```
It might be a config, or a function that returns void, config, promise<void>, promise<config>. Might be easiest to skip typing.

davismj replied:
@bigopon

davismj commented on src/router.ts:
```diff
@@ -278,6 +329,7 @@ export class Router {
       throw new Error(`A route with name '${name}' could not be found. Check that \`name: '${name}'\` was specified in the route's config.`);
     }
 
+    debugger;
```
rip

bigopon replied:
lol

davismj commented on src/router.ts:
```diff
@@ -289,7 +341,12 @@ export class Router {
+      this,
+      'href' in config
+        ? config.href
+        // potential error when config.route is a string[] ?
```
looks like it to me too

bigopon replied:
Maybe this ?

https://github.com/aurelia/router/commit/0cd5423a421c17d6bf0b10598ea451902ec5a590#diff-15ee858ae5d5aa9b3e2846cfd1c4c466R303

But it still feels a bit awkward to me, maybe we need to improve this a bit

davismj commented on src/router.ts:
```diff
@@ -491,7 +556,9 @@ export class Router {
-        queryParams: first.queryParams || results.queryParams,
+        queryParams: first.queryParams
+          // How is this valid
+          || (results as any).queryParams,
```
It's kind of a funny class. It doesn't inherit from Array but it adopts some of the Array methods right off the prototype so it looks and feels like an array but isn't.

Assuming this is typed correctly above, you shouldn't need the `as any`

https://github.com/aurelia/route-recognizer/blob/master/src/route-recognizer.js#L252

davismj replied:
it is mistyped in route recognizer. do (as RecognizeResults) for now.

davismj commented on src/router.ts:
```diff
@@ -505,7 +572,10 @@ export class Router {
-        queryParams: results ? results.queryParams : {},
+        queryParams: results
+          // how is this valid
+          ? (results as any).queryParams
```
(as RecognizeResults)

davismj commented on src/router.ts:
```diff
@@ -518,7 +588,10 @@ export class Router {
-          queryParams: results ? results.queryParams : {},
+          queryParams: results
+            // how is this valid
+            ? (results as any).queryParams
```
(as RecognizeResults)

bigopon commented on src/router.ts:
```diff
@@ -366,7 +427,8 @@ export class Router {
 
       this.navigation.push(navModel);
-      this.navigation = this.navigation.sort((a, b) => a.order - b.order);
+      // this is a potential error / inconsistency between browsers
```
I was unclear here. Only returning `0` will cause inconsistency behavior, per spec. From MDN:

> If compareFunction(a, b) returns 0, leave a and b unchanged with respect to each other, but sorted with respect to all different elements. Note: the ECMAscript standard does not guarantee this behaviour, and thus not all browsers (e.g. Mozilla versions dating back to at least 2003) respect this.

Because of that, if we have a list of route configs that all have `nav: true`, it will fall in this `return 0` behavior, thus may yield inconsistency

bigopon commented:
@davismj 

https://github.com/aurelia/router/pull/615/files#diff-52fd7ff9c4a70144c285b45938c2af80L44

The above block starting from that line seems to be wrong to me. It should be changed

from:

```js
let viewModel = infos[i];
let result = viewModel[callbackName](navigationInstruction);
return processPotential(result, inspect, next.cancel);
```

to

```js
let current= infos[i];
let result = current.viewModel[callbackName](navigationInstruction);
return processPotential(result, inspect, next.cancel);
```

Similar to this block

https://github.com/aurelia/router/pull/615/files#diff-52fd7ff9c4a70144c285b45938c2af80L121

davismj commented on src/pipeline-provider.ts:
```diff
@@ -110,7 +126,8 @@ export class PipelineProvider {
 
-  _createPipelineSlot(name, alias) {
+  /**@internal */
+  _createPipelineSlot(name: string, alias?: string) {
```
This always returns a PipelineSlot

bigopon commented:
@davismj I've fixed most of the issues except this
```ts
steps: (Function | PipelineSlot)[];
```
Right in the constructor of it, we see how `Function` is a valid type for it.

bigopon commented:
@davismj I've merged all changes in master into my PR branch. It's good for final review / merge.

bigopon commented:
@davismj I've merged all changes in master into my PR branch. It's good for final review / merge.
cc @EisenbergEffect @fkleuver

bigopon commented:
The interface of `component` in view port has become strict, this may, or may not be an issue. If folks complain, we can turn  it into an index-able interface to avoid typing issue.
